### PR TITLE
New version of the script_score term stats helpers.

### PR DESCRIPTION
**Pull request by afoucret**
A new version of the script_score term stats helpers.

## Overview

This PR exposes term statistics for the child query through the `_termStatistics` variable (instead of having parameters to specify the field and query as parameters of the helper).

As an example, the following query: 

```
{
  "query": {
    "script_score": {
      "query": { "match": { "title": "star wars" } },
      "script": {
        "source": "return _termStatistics.docFreq().getSum()"
      }
    }
  }
}
```

will read document frequency for terms `star` and `wars` in the title field and return their sum.

The following method are available in the `_termStatistics` object:

long uniqueTermsCount() | Number of unique terms in the query
-- | --
long matchedTermsCount() | Number of terms of the query matched in the current document
DoubleSummaryStatistics docFreq() | Statistics for the document frequency of the terms (number of documents in the index with the term).
DoubleSummaryStatistics totalTermFreq() | Statistics for the total term frequency of the terms (number of occurrence for a term in the index with the term).
DoubleSummaryStatistics termFreq() | Statistics for the term frequency of the terms (number of occurrence of the term in the current document)
DoubleSummaryStatistics termPositions() | Statistics for the term position of the terms in the current document

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java
Comment (ioanatia) [inline server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java:89]: when is this branch executed? and do we still need the `needsTermStatistics` check here and do `this.visit(QueryVisitor.termCollector(terms));`?

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java
Comment (ioanatia) [inline server/src/main/java/org/elasticsearch/script/ScoreScript.java:98]: completely unrelated to your changes, but I guess this line is not needed, no?

File: modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.score.txt
Comment (afoucret) [inline modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.score.txt:19]: TODO : remove this cause we do not want to expose it

File: modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.score.txt
Comment (afoucret) [inline modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.score.txt:22]: TODO : remove this cause we do not want to expose it

**Review by jimczi — COMMENTED** [review]
The approach looks good to me, it is isolated and script_score queries that don't need the term statistics won't be impacted so +1.

@rjernst you had some concerns regarding the re-introduction of term stats in script, would this approach alleviates those?

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/190_term_statistics_script_score.yml
Comment (afoucret) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/190_term_statistics_script_score.yml:1]: ℹ️ YAML rest test for many use case.

**Comment by afoucret** [comment]
@elastic/es-core-infra would be great if someone could review this PR

File: server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java
Comment (afoucret) [inline server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java:94]: ℹ️ Required so termStats are collected during DFS when `needs_termStatistics` is true.

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java
Comment (afoucret) [inline server/src/main/java/org/elasticsearch/script/ScoreScript.java:222]: ℹ️ This methods is automatically implemented by the painless script factory and return `true` if the `_termStatistics` variable is used in the script.

**Review by rjernst — COMMENTED** [review]
@afoucret My concerns of performance hits from reintroducing term stats are mostly allayed here. This PR is well put together, thank you!

Most of my comments are small. One larger concern is over the caching strategy. I think we should avoid maps. Although using HashMap has constant time lookup, we still must compute the hash N * T (number of documents scored, number of terms in query) times. Unlike the previous term stats API in scripts here we do not give access to per term stats (which is a great simplification!). But then we don't need the associative lookup, just the ability to iterate over the terms enums/states.

Additionally, I think we should have unit tests for each of the methods on ScriptTermStats (when retrieved from a ScoreScript), making sure to hit all edge cases handled in these methods.

Finally, I have a very minor bikeshed about the naming. I see you mimicked the underscore prefix of score. Note that `_score` is somewhat a historical name that we've been stuck with. And since we don't restrict variable names from users, underscore isn't actually a protected namespace (though maybe it should be!). One concrete feedback we've received in the past with painless is long names. So I wonder if it should just be `termStats` (or `_termStats`).

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:116]: Done.

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:237]: When does a IAE throw here? It seems weird to just ignore it.

**Review by benwtrent — COMMENTED** [review]
Can we have a test similar to `NeedsScoreTests`?

Overall, this looks good. I do wonder about the default `leafFactor` impl as Ryan does.

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/ScriptTermStats.java]: I introduced a new `StatsAccumulator` class that does more or less the same than `DoubleSummaryStatistics` but that can be reset.

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/ScriptTermStats.java]: The code has been refactored and should looks like what you described. Is it better right now?

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/ScoreScript.java]: I wanted to avoid reimplementing every time we have a anonymous class inheriting from ScriptScore (there is a lot of them in the tests). But I changed my mind since it is probably weird to have a default implementation for it.

File: server/src/test/java/org/elasticsearch/search/query/ScriptScoreQueryTests.java
Comment (afoucret) [reply server/src/test/java/org/elasticsearch/search/query/ScriptScoreQueryTests.java]: I did implement two tests cases to check if the termStats is available only if `needs_termStats` is true

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/ScriptTermStats.java]: Indeed, it is never the case. I removed the conditional for it.

**Comment by afoucret** [comment]
@rjernst @benwtrent, I think all the comments from your reviews have been addressed through the refactoring and adding a lot of tests. I would be glad to read your review again.

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:30]: This could be `IntSupplier`

**Review by rjernst — COMMENTED** [review]
This looks really close. I have a few more thoughts.

File: server/src/main/java/org/elasticsearch/script/StatsAccumulator.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/StatsAccumulator.java]: Renamed into `SfatsSummary` per your suggestion.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/190_term_statistics_script_score.yml
Comment (afoucret) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/190_term_statistics_script_score.yml]: I introduced a new feature instead.

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:131]: `termStatistics` is using the `Searcher::termStatistics` method:

```
private TermStatistics termStatistics(Term term, TermStates termStates) throws IOException {
    return searcher.termStatistics(term, termStates.docFreq(), termStates.totalTermFreq());
}
When a term is not present in the current shards, this is equivalent to:

searcher.termStatistics(term, 0, 0);
which uses aggregated Document Frequency Statistics (DFS) to retrieve term statistics:

if (aggregatedDfs == null) {
    // Either the DFS phase is executing, or the search type does not include a DFS phase.
    return super.termStatistics(term, docFreq, totalTermFreq);
}

TermStatistics termStatistics = aggregatedDfs.termStatistics().get(term);
if (termStatistics == null) {
    // No stats available for this term - it could be from a "must_not" clause or similar, 
    // preventing term extraction from the query.
    return super.termStatistics(term, docFreq, totalTermFreq);
}

return termStatistics;
```

when a term is missing it will try to create a `TermStatistics` object with a docFreq  equal to 0, causing the IAE to be raised.

Because we want to leverage the stats from dfs for more reliable results in the distributed use case, we have to catch this edge case since there is no way to predict when the IAE will be thrown.

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:131]: This seems fragile. There could be other cases (for example, an actual bad argument, like reversing term freq and total term freq) for which an IAE is thrown. I think we should find a way to avoid calling termStatistics. For example, you could have a new method on ContextIndexSearcher, to find the totalTermFreq for a term (which takes a fallback). So something like:

```
if (searcher instanceof ContextIndexSearcher dfsSearcher) {
    totaltermFreqStats.accept(dfsSearcher.getTotalTermFreq(term, totalTermFreq));
} else {
    totaltermFreqStats.accept(totalTermFreq);
}
```

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/ScoreScript.java:206]: `_termStats` is currently implemented in the `script_score` query only.

While we want to implement it in some other place (function_score), it is not yet the case (should be easy to do once this PR will be merged). The exception makes it clear to the user that `_termStats` is not available in that case.

File: test/framework/src/main/java/org/elasticsearch/script/MockScriptEngine.java
Comment (afoucret) [reply test/framework/src/main/java/org/elasticsearch/script/MockScriptEngine.java:744]: I prefer to keep false since we would have to implement a mock for ScriptTermStats in that case.
BTW, there is specific tests for the termStats use case.

**Comment by afoucret** [comment]
@rjernst I addressed most of your comment and feedback. It would be great if you have the time for another review (🤞 and maybe an approve this PR!!!!)

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/ScriptTermStats.java]: I like your proposal and implemented something very similar in the last version I just pushed.

File: server/src/main/java/org/elasticsearch/script/ScoreScript.java
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/ScoreScript.java]: I did update it to an assertion per your recommendation.

**Review by rjernst — APPROVED** [review]
Thanks for all the iterations. LGTM.

**Pull request merged**
